compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3747,6 +3784,30 @@ private boolean isAssignableToUnionType(BType source, BType target, Set<TypePair
         return true;
     }
 
+    private boolean isUnionDistinctError(Set<BType> memberTypesSet) {
+        List<BType> memberTypes = new ArrayList<>(memberTypesSet);
+        int size = memberTypes.size();
+        for (int i = 0; i < size - 1; i++) {
+            BErrorType t1 = (BErrorType) memberTypes.get(i);
+            BErrorType t2 = (BErrorType) memberTypes.get(i + 1);
+            Set<BTypeIdSet.BTypeId> t1TypeIds = t1.typeIdSet.getAll();
```

Comment:
If we cache the first member, we can get his type id set as a final set and cache it outside for loop also and compare again others. We don't need to call `getAll` many times and also create many temporary objects.
As an example for a 4 element set, this logic will call `getAll` 6 times with duplicate calls. But if we cache first one we will call it only 4 times.

Reply:
(e1, e2, e3, e4)
iteration 1 : `e1.getAll, e2.getAll`
iteration 2 : `e2.getAll, e3.getAll`
iteration 3 : `e3.getAll, e4.getAll`

As you can see ` e2.getAll` and `e3.getAll` is repeated.

Reply:
Changed